services/autoscaling/src/simcore_service_autoscaling/modules/instrumentation.py

```diff
@@ -88,29 +159,130 @@ def __post_init__(self) -> None:
+            "terminated_instances_total",
             "Number of EC2 instances that were terminated",
             labelnames=EC2_INSTANCE_LABELS,
             namespace=METRICS_NAMESPACE,
             subsystem=self.subsystem,
         )
 
     def update_from_cluster(self, cluster: Cluster) -> None:
-        _update_gauge(self._active_nodes, cluster.active_nodes)
-        _update_gauge(self._pending_nodes, cluster.pending_nodes)
-        _update_gauge(self._drained_nodes, cluster.drained_nodes)
-        _update_gauge(self._buffer_drained_nodes, cluster.reserve_drained_nodes)
-        _update_gauge(self._pending_ec2s, cluster.pending_ec2s)
-        self._disconnected_nodes.set(len(cluster.disconnected_nodes))
+        self._cluster_metrics.update_from_cluster(cluster)
 
     def instance_started(self, instance_type: str) -> None:
         self._started_instances.labels(instance_type=instance_type).inc()
 
+    def instance_launched(self, instance_type: str) -> None:
+        self._launched_instances.labels(instance_type=instance_type).inc()
+
+    def instance_stopped(self, instance_type: str) -> None:
+        self._stopped_instances.labels(instance_type=instance_type).inc()
+
     def instance_terminated(self, instance_type: str) -> None:
         self._terminated_instances.labels(instance_type=instance_type).inc()
 
 
+P = ParamSpec("P")
+R = TypeVar("R")
+
+
+def _instrumented_ec2_client_method(
+    metrics_handler: Callable[[str], None],
+    *,
+    instance_type_from_method_arguments: Callable[..., list[str]] | None,
+    instance_type_from_method_return: Callable[..., list[str]] | None,
+) -> Callable[
+    [Callable[P, Coroutine[Any, Any, R]]],
+    Callable[P, Coroutine[Any, Any, R]],
+]:
+    def decorator(
+        func: Callable[P, Coroutine[Any, Any, R]]
+    ) -> Callable[P, Coroutine[Any, Any, R]]:
+        @functools.wraps(func)
+        async def wrapper(*args: P.args, **kwargs: P.kwargs) -> R:
+            result = await func(*args, **kwargs)
+            if instance_type_from_method_arguments:
+                for instance_type in instance_type_from_method_arguments(
+                    *args, **kwargs
+                ):
+                    metrics_handler(instance_type)
+            elif instance_type_from_method_return:
+                for instance_type in instance_type_from_method_return(result):
+                    metrics_handler(instance_type)
+            return result
+
+        return wrapper
+
+    return decorator
+
+
+def _get_instance_type_from_launch_instances_return_value(
+    instance_datas: list[EC2InstanceData],
+) -> list[str]:
+    return [i.type for i in instance_datas]
+
+
+def _get_instance_type_from_stop_instances_args(
+    instance_datas: Iterable[EC2InstanceData],
+) -> list[str]:
+    return [i.type for i in instance_datas]
+
+
+def _get_instance_type_from_terminate_instances_args(
+    instance_datas: Iterable[EC2InstanceData],
+) -> list[str]:
+    return [i.type for i in instance_datas]
```

Comment:
the code here looks the same to me, am I missing something? Can't we have them use some common function?

Reply:
yes you are partially right. will check what I can do